proof/verifier.go

```diff
@@ -180,71 +226,276 @@ func (p *Proof) verifySplitRootProof() error {
+func (p *Proof) verifyV2ExclusionProofs(
+	p2trOutputs map[uint32]fn.Set[asset.SerializedKey]) (
+	*commitment.TapCommitmentVersion, error) {
+
+	commitVersions := make(map[uint32][]commitment.TapCommitmentVersion)
+
+	// Collect the STXOs from the new asset.
+	stxoAssets, err := asset.CollectSTXO(&p.Asset)
+	if err != nil {
+		return nil, fmt.Errorf("error collecting STXO assets: %w", err)
+	}
+
+	// Map STXOs by serialized key.
+	assetMap := make(map[asset.SerializedKey]*asset.Asset)
+	for idx := range stxoAssets {
+		stxoAsset := stxoAssets[idx].(*asset.Asset)
+		key := asset.ToSerialized(stxoAsset.ScriptKey.PubKey)
+		assetMap[key] = stxoAsset
+		for outIdx := range p2trOutputs {
+			p2trOutputs[outIdx].Add(key)
+		}
+	}
+
+	for idx := range p.ExclusionProofs {
+		baseProof := p.ExclusionProofs[idx]
+		if baseProof.TapscriptProof != nil &&
+			baseProof.TapscriptProof.Bip86 {
+
+			// This is a BIP-0086 wallet output (likely a change
+			// output) that does not commit to any script or Taproot
+			// Asset root. There is no special stxo proof handling
+			// needed, so we default to `handleBasicExclusionProof`.
+			err := p.handleBasicExclusionProof(
+				&baseProof, p2trOutputs, commitVersions,
+			)
+			if err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		_, err := p.verifySTXOProofSet(
+			baseProof, assetMap, p2trOutputs, commitVersions, false,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = p.verifyRemainingOutputs(p2trOutputs)
+	if err != nil {
+		return nil, err
+	}
+
 	if len(commitVersions) == 0 {
 		return nil, nil
 	}
 
-	// All ExclusionProofs must have similar versions.
-	firstCommitVersion := maps.Values(commitVersions)[0]
-	for outputIdx, commitVersion := range commitVersions {
-		outputCommitVersion := commitVersion
-		if !commitment.IsSimilarTapCommitmentVersion(
-			&firstCommitVersion, &outputCommitVersion,
-		) {
+	// All proofs must have similar versions.
+	firstVersion := maps.Values(commitVersions)[0][0]
+	return p.verifySTXOVersions(firstVersion, commitVersions)
+}
+
+// handleBasicExclusionProof handles a basic (non-STXO) exclusion proof.
+func (p *Proof) handleBasicExclusionProof(baseProof *TaprootProof,
+	p2trOutputs map[uint32]fn.Set[asset.SerializedKey],
+	commitVersions map[uint32][]commitment.TapCommitmentVersion) error {
+
+	derivedCommitment, err := verifyTaprootProof(
+		&p.AnchorTx, baseProof, &p.Asset, false,
+	)
+	if err != nil {
+		return fmt.Errorf("error verifying STXO proof: %w", err)
+	}
+
+	outputIdx := baseProof.OutputIndex
+	delete(p2trOutputs, baseProof.OutputIndex)
+
+	// Store commitment version if Taproot Asset commitment present.
+	if derivedCommitment != nil {
+		commitVersions[outputIdx] = append(
+			commitVersions[outputIdx],
+			derivedCommitment.Version,
+		)
+	}
+
+	return nil
+}
+
+// verifySTXOProofSet verifies a set of STXO proofs.
+func (p *Proof) verifySTXOProofSet(baseProof TaprootProof,
```

Comment:
Could probably pass in the anchor TX here and make this a pure function as well.